### DISCUSSION TITLE
Connection.upsert for Inserting New Records Using Id as the External ID

### DISCUSSION THREAD

**Discussion by tarot** [jsforce/jsforce#1489] [Ideas]
I encountered an error `METHOD_NOT_ALLOWED: HTTP Method 'PATCH' not allowed. Allowed are GET,HEAD,POST` when `connection.sobject('Account').upsert({Name: 'foo'}, 'Id')`.

I found description below in developer guide.
https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/dome_upsert.htm

> Inserting New Records Using Id as the External ID
> This example uses the POST method as a special case to insert a record where the Id field is treated as the external ID. Because the value of Id is null, it’s omitted from the request. This pattern is useful when you’re writing code to upsert multiple records by different external IDs and you don’t want to request a separate resource. POST using Id is available in API version 37.0 and later.

So if external ID field is Id and Id is not present, HTTP method for upsert should be `POST`.

These lines:
https://github.com/jsforce/jsforce/blob/a99a38b55346388f53a47d8b2e5afcace43c232b/lib/connection.js#L754

```javascript
      var url = [ self._baseUrl(), "sobjects", sobjectType, extIdField, extId ].join('/');
      return self.request({
        method : 'PATCH',
        url : url,
```

should be like:

```javascript
      var url = _.compact([ self._baseUrl(), "sobjects", sobjectType, extIdField, extId ]).join('/');
      return self.request({
        method : (extIdField.toUpperCase() === 'ID' && extId == null) ? 'POST' : 'PATCH',
        url : url,
```

**Comment by tamarklg** [comment]
@stomita 
I came across this problem, too.
Does someone have any solution?

**Comment by stomita** [comment]
The `upsert` in jsforce is not supporting the ID based upsert. How about using `create()` if you surely want to insert the new records ? (Personally I'm not sure why these special upsert is existing  in the API)

**Comment by trueter** [comment]
I don't have the code anymore, but I remember monkey patching the internals of jsforce to add this functionality wasn't very hard.

**Comment by paustint** [comment]
The bulk API supports this use-case.

**Comment by bsansone** [comment]
@stomita Hi, this is the current use-case I am looking to use `upsert` for:

In my api, I am making a request to update a master object in SalesForce, when this request is made, there may or may not exist a related detail object to this master object. I determine this by making a query to check for that detail object where the `Master-Detail` field is equal to the master object `Id`, e.g. `SELECT Id FROM DetailObject__c WHERE MasterDetailField__c = 'MasterObjectId';` If no records are returned I am creating that detail object, if records are returned than I know I can update the detail object.

So instead of conditionally checking if there are records returned myself and either making a `create` or `update` call, I wanted to use `upsert` to do it for me, using @tarot 's suggested change, I was able to get this to work:

```js
const record = {
    Id: detailObjectRecordExists ? detailObjectId : null,
    Field1__c: "foo",
    Field2__c: "bar"
};

if (!detailObjectRecordExists) {
    record.MasterDetailField__c = MasterObjectId;
}

conn.sobject("DetailObject__c").upsert(record, "Id", function(err, ret) {
    // ...
});
```

**Comment by stomita** [comment]
@bsansone OK with your use case which totally may happen.

However still I cannot understand why the API is existing, it still requires checking and switching to call proper http method internally in the`upsert()` method. It seems for me that the special interface for Id as external ID upsert is not so much needed as we can call the usual POST endpoint for record creation when switching.

Anyway, at least JSforce library user would get benefit without thinking the switch, it is worth considering implementation.

**Comment by omniphx** [comment]
Is it possible to simply change the HTTP method from PATCH to POST in this scenario?
```
const result: RecordResult = await salesforce.sobject('FieldPermissions').upsert(fieldPermissionsToSaveValues, 'Id', {method:POST});
```

**Comment by edelauna** [comment]
Your payload is missing the external field as such the request URL is breaking - I submitted [PR#931 ](https://github.com/jsforce/jsforce/pull/931)

In the mean time try adding a slash to the extIdField field so that it reads:

`connection.sobject('Account').upsert({Name: 'foo'}, 'Id/')`

or adding Id to your record (doesn't work if Id is `null` or `" "` unfortunately). 

`connection.sobject('Account').upsert({Name: 'foo', Id:'123'}, 'Id/')`
